File: host/examples/vector_fma.c
/* Copyright 2026 Logan W.
 * SPDX-License-Identifier: Apache-2.0
 *
 * The smallest useful libcft program: run a fused multiply-add over a
 * vector at every precision and print a checksum of the result.
 *
 *     cc -Ihost/include host/examples/vector_fma.c host/libcft.a -o vector-fma
 *     ./vector-fma
 *
 * The checksum is the point. Run this on another machine, in another
 * operating system, built by another compiler, and the lines must be
 * character for character identical - and identical again to what
 * host/examples/vector_fma_ctypes.py prints, which reaches the same
 * library through Python instead of C. That is the entire product
 * reduced to something you can diff.
 *
 * Nothing here needs a card. Pass an xclbin path as argv[1] and the
 * same program runs on the tile instead, with the same output.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "cft.h"

#define N 4096

/* ---- the operand stream ------------------------------------------ *
 *
 * xorshift32, stepped once per byte, so that the C and Python programs
 * can generate byte-identical inputs without sharing a file. It is not
 * a good random number generator and does not need to be: it needs to
 * be short enough to reimplement without error in any language, which
 * a good one would not be.
 */
static uint32_t rng_state;

static void rng_seed(uint32_t s)
{
    rng_state = s ? s : 1u;
}

static uint8_t rng_byte(void)
{
    rng_state ^= rng_state << 13;
    rng_state ^= rng_state >> 17;
    rng_state ^= rng_state << 5;
    return (uint8_t)(rng_state & 0xffu);
}

/* One element: a normal with a fraction from the stream and an
 * exponent within 32 of 1.0, so that products stay in range and the
 * run exercises rounding rather than overflow. */
static void make_element(uint8_t *e, int width, int exp_w, int man_w, int bias)
{
    int nb = width / 8, kb = man_w / 8, rb = man_w % 8, j;
    uint32_t ef;

    for (j = 0; j < nb; j++)
        e[j] = rng_byte();
    ef = (uint32_t)(bias - 32 + (int)(rng_byte() & 63u));

    e[kb] &= (uint8_t)((1u << rb) - 1u);
    for (j = kb + 1; j < nb; j++)
        e[j] = 0;
    for (j = 0; j < exp_w; j++)
        if ((ef >> j) & 1u)
            e[(man_w + j) / 8] |= (uint8_t)(1u << ((man_w + j) % 8));
    if (rng_byte() & 1u)
        e[nb - 1] |= 0x80u;
}

/* FNV-1a, 64-bit. The constants are written in hex because that is
 * how they are specified; transcribing them as decimal is how you get
 * a checksum that is stable, plausible, and not FNV. */
static uint64_t fnv1a(const uint8_t *p, size_t n)
{
    uint64_t h = 0xcbf29ce484222325ull;
    size_t i;
    for (i = 0; i < n; i++) {
        h ^= p[i];
        h *= 0x100000001b3ull;
    }
    return h;
}

struct geom { int width, exp_w, man_w, bias; };

int main(int argc, char **argv)
{
    static const struct geom g[4] = {
        {  32,  8,  23,    127 },
        {  64, 11,  52,   1023 },
        { 128, 15, 112,  16383 },
        { 256, 19, 236, 262143 }
    };
    const char *artifact = (argc > 1) ? argv[1] : NULL;
    cft_device *dev = NULL;
    cft_status st;
    int f;

    st = cft_open(artifact, 0, &dev);
    if (st != CFT_OK) {
        fprintf(stderr, "cft_open(%s): %s\n",
                artifact ? artifact : "software", cft_strerror(st));
        return 1;
    }
    printf("cft-fp256 vector fma, %s backend\n",
           artifact ? artifact : "software");

    for (f = 0; f < 4; f++) {
        size_t esz = cft_format_size((cft_format)f);
        uint8_t *a, *b, *c, *d;
        uint32_t flags = 0;
        size_t i;

        if (!cft_supports(dev, CFT_FMA, (cft_format)f)) {
            printf("%-6s not available on this device\n",
                   cft_format_name((cft_format)f));
            continue;
        }

        a = (uint8_t *)malloc(N * esz);
        b = (uint8_t *)malloc(N * esz);
        c = (uint8_t *)malloc(N * esz);
        d = (uint8_t *)malloc(N * esz);
        if (!a || !b || !c || !d) {
            fprintf(stderr, "out of memory\n");
            free(a); free(b); free(c); free(d);
            cft_close(dev);
            return 1;
        }

        /* Seeded per format, so each line is independent of the ones
         * before it and can be reproduced on its own. */
        rng_seed(0x1234567u + (uint32_t)f);
        for (i = 0; i < N; i++) {
            make_element(a + i * esz, g[f].width, g[f].exp_w, g[f].man_w,
                         g[f].bias);
            make_element(b + i * esz, g[f].width, g[f].exp_w, g[f].man_w,
                         g[f].bias);
            make_element(c + i * esz, g[f].width, g[f].exp_w, g[f].man_w,
                         g[f].bias);
        }

        st = cft_run(dev, CFT_FMA, (cft_format)f, CFT_RNE,
                     a, b, c, d, N, &flags, NULL);
        if (st != CFT_OK) {
            fprintf(stderr, "cft_run %s: %s\n",
                    cft_format_name((cft_format)f), cft_strerror(st));
            free(a); free(b); free(c); free(d);
            cft_close(dev);
            return 1;
        }

        printf("%-6s n=%d rne  checksum 0x%016llx  flags 0x%02x\n",
               cft_format_name((cft_format)f), N,
               (unsigned long long)fnv1a(d, N * esz), (unsigned)flags);

        free(a); free(b); free(c); free(d);
    }

    cft_close(dev);
    return 0;
}

File: host/examples/vector_fma.jl
# Copyright 2026 Logan W.
# SPDX-License-Identifier: Apache-2.0
#
# The same program as vector_fma.c, in Julia, through ccall.
#
#     julia host/examples/vector_fma.jl [artifact.xclbin]
#
# It must print exactly what the C program prints - every line, every
# digit. Nothing beyond the standard library is used: Libdl finds the
# shared library, Printf formats the lines, and ccall is part of the
# language. No package, no Project.toml, no binding generator - the
# claim cft.h makes about its ABI, paid out in another language.
#
# As in the ctypes port, the arithmetic is not reimplemented here.
# Only the operand stream and the checksum are, because they are the
# test harness rather than the product; everything the library
# guarantees still happens inside the library.
#
# *** UNVERIFIED ***
# The machine this file was written on has no Julia toolchain - none
# on PATH, none in the usual install directories, none in the repo's
# containers - so unlike every other file in examples/ this one has
# not been run against the library it binds. The claim it makes is
# still machine-checkable: `make -C host examples-lang` diffs its
# output against the C example's on any machine that does have julia.
# Run that before trusting this file, and delete this block when it
# passes - an example should not keep its asterisk one commit longer
# than the asterisk is true.

using Libdl
using Printf

# cft_op and cft_round - the values are normative (cft.h), so a
# caller writes the names and never the register layout behind them.
const FMA = Cint(0)
const RNE = Cint(0)

const N = 4096

# width, exp_w, man_w, bias - the geometry table vector_fma.c
# carries. This is operand-stream machinery, not format knowledge the
# caller should own: element sizes and format names below still come
# from the library, because a size the caller guesses is a size that
# goes stale.
const GEOM = [
    (32, 8, 23, 127),
    (64, 11, 52, 1023),
    (128, 15, 112, 16383),
    (256, 19, 236, 262143),
]

# Exactly one candidate, chosen by platform - the same policy as
# vector_fma_ctypes.py, for the same reason: listing every name and
# taking the first that exists looks tolerant and is not. A tree
# built from two platforms holds both libraries, and the fallback
# then loads the foreign one and fails with a loader error instead of
# "build it first".
function library_path()
    override = get(ENV, "CFT_LIB", "")
    isempty(override) || return override
    name = Sys.iswindows() ? "cft.dll" :
           Sys.isapple() ? "libcft.dylib" : "libcft.so"
    root = normpath(joinpath(@__DIR__, "..", ".."))
    return joinpath(root, "host", name)
end

function load()
    path = library_path()
    if !isfile(path)
        println(stderr, "$path not found - run `make -C host` first")
        exit(1)
    end
    return dlopen(path)
end

# xorshift32, stepped once per byte - the stream vector_fma.c
# generates, which is what lets the two checksums be compared at all.
# Julia's fixed-width unsigned integers shift and wrap exactly as C's
# unsigned ints do, so the port is line for line.
mutable struct Rng
    s::UInt32
    Rng(seed::Integer) = new(seed == 0 ? UInt32(1) : UInt32(seed))
end

function rng_byte!(r::Rng)
    s = r.s
    s = xor(s, s << 13)
    s = xor(s, s >> 17)
    s = xor(s, s << 5)
    r.s = s
    return UInt8(s & 0xff)
end

# One element at byte offset `off` of e: a normal with a fraction
# from the stream and an exponent within 32 of 1.0, so that products
# stay in range and the run exercises rounding rather than overflow.
# vector_fma.c line for line, shifted to 1-based indexing - the
# offsets are the one place a Julia port can silently diverge, which
# is precisely what the checksum diff exists to catch.
function make_element!(e::Vector{UInt8}, off::Int, r::Rng,
                       width::Int, exp_w::Int, man_w::Int, bias::Int)
    nb = div(width, 8)
    kb = div(man_w, 8)
    rb = man_w % 8

    for j in 1:nb
        e[off + j] = rng_byte!(r)
    end
    ef = bias - 32 + Int(rng_byte!(r) & 0x3f)

    e[off + kb + 1] &= UInt8((1 << rb) - 1)
    for j in (kb + 2):nb
        e[off + j] = 0x00
    end
    for j in 0:(exp_w - 1)
        if ((ef >> j) & 1) != 0
            bit = man_w + j
            e[off + div(bit, 8) + 1] |= UInt8(1) << (bit % 8)
        end
    end
    if (rng_byte!(r) & 0x01) != 0x00
        e[off + nb] |= 0x80
    end
end

# FNV-1a, 64-bit. The constants are written in hex because that is
# how they are specified; transcribing them as decimal is how you get
# a checksum that is stable, plausible, and not FNV.
function fnv1a(p::Vector{UInt8})
    h = 0xcbf29ce484222325
    for b in p
        h = xor(h, b)
        h *= 0x100000001b3
    end
    return h
end

function main()
    artifact = isempty(ARGS) ? nothing : ARGS[1]
    backend = artifact === nothing ? "software" : artifact
    lib = load()

    # One dlsym per entry point. ccall through a looked-up pointer is
    # the stdlib route to a library whose location is decided at run
    # time; the tuple of argument types is the entire binding, which
    # is the point being demonstrated.
    cft_open        = dlsym(lib, :cft_open)
    cft_close       = dlsym(lib, :cft_close)
    cft_run         = dlsym(lib, :cft_run)
    cft_supports    = dlsym(lib, :cft_supports)
    cft_format_size = dlsym(lib, :cft_format_size)
    cft_format_name = dlsym(lib, :cft_format_name)
    cft_strerror    = dlsym(lib, :cft_strerror)

    strerror(st) =
        unsafe_string(ccall(cft_strerror, Ptr{UInt8}, (Cint,), st))

    dev = Ref{Ptr{Cvoid}}(C_NULL)
    st = ccall(cft_open, Cint, (Ptr{UInt8}, Cint, Ptr{Ptr{Cvoid}}),
               artifact === nothing ? C_NULL : artifact, Cint(0), dev)
    if st != 0
        println(stderr, "cft_open($backend): $(strerror(st))")
        exit(1)
    end
    println("cft-fp256 vector fma, $backend backend")

    for f in 0:3
        (width, exp_w, man_w, bias) = GEOM[f + 1]
        fmt = Cint(f)

        # The name comes from the library, not a local table - cft.h
        # provides cft_format_name precisely so a binding, a log line
        # and a conformance report all say the same thing.
        name = unsafe_string(ccall(cft_format_name, Ptr{UInt8}, (Cint,),
                                   fmt))

        if ccall(cft_supports, Cint, (Ptr{Cvoid}, Cint, Cint),
                 dev[], FMA, fmt) == 0
            @printf("%-6s not available on this device\n", name)
            continue
        end

        esz = Int(ccall(cft_format_size, Csize_t, (Cint,), fmt))

        a = Vector{UInt8}(undef, N * esz)
        b = Vector{UInt8}(undef, N * esz)
        c = Vector{UInt8}(undef, N * esz)
        d = Vector{UInt8}(undef, N * esz)

        # Seeded per format, so each line is independent of the ones
        # before it and can be reproduced on its own.
        r = Rng(0x1234567 + f)
        for i in 0:(N - 1)
            make_element!(a, i * esz, r, width, exp_w, man_w, bias)
            make_element!(b, i * esz, r, width, exp_w, man_w, bias)
            make_element!(c, i * esz, r, width, exp_w, man_w, bias)
        end

        flags = Ref{UInt32}(0)
        st = ccall(cft_run, Cint,
                   (Ptr{Cvoid}, Cint, Cint, Cint,
                    Ptr{UInt8}, Ptr{UInt8}, Ptr{UInt8}, Ptr{UInt8},
                    Csize_t, Ptr{UInt32}, Ptr{UInt32}),
                   dev[], FMA, fmt, RNE, a, b, c, d, N, flags, C_NULL)
        if st != 0
            println(stderr, "cft_run $name: $(strerror(st))")
            ccall(cft_close, Cvoid, (Ptr{Cvoid},), dev[])
            exit(1)
        end

        @printf("%-6s n=%d rne  checksum 0x%016x  flags 0x%02x\n",
                name, N, fnv1a(d), flags[])
    end

    ccall(cft_close, Cvoid, (Ptr{Cvoid},), dev[])
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
